templates/README.md

```diff
@@ -0,0 +1,276 @@
+# Template development guidelines
+
+The idea is to reuse and enhance the concept of OpenShift templates and its parameter expansion and also to reuse the kubevirt’s VM.spec structure as a DTO that will provide data to the parameter expansion.
+
+Since the templates will contain lots of redundancies, there should be a build time templating mechanism for composing the final templates from snippets.
+
+Also, there might be multiple different OSes, flavors, sizes mentioned by any single template file if they share the same configuration.
+
+Please note that the cnv.io suffix used in labels and annotations is temporary and is likely to change.
+
+## User Experience
+
+The flow should be the same as usual with regards to openshift templates. The user either takes a local template and generates a VM object.
+
+Or the user gets a template from kubernetes and invokes the proper REST command on top of it to keep the VM linked to the template or edit it in accordance to the template.
+
+The UI can then operate on top of template parameters and/or the Template.VM.spec DOM directly. There is an openshift [REST endpoint for instantiating Templates](https://docs.openshift.com/container-platform/3.7/rest_api/examples.html#template-instantiation) and the UI might be able to use it to create an in memory instance of VM it will then work with before posting it (kubernetes dry-run issues [https://github.com/kubernetes/kubernetes/issues/11488](#11488) and [https://github.com/kubernetes/kubernetes/issues/63559](#63559)).
+
+## Example file
+
+### template.yaml
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: v1
+kind: Template # Openshift kind of template or something "better"
+metadata:
+  name: windows-10
+  annotations:
+    openshift.io/display-name: "Generic Microsoft Windows 7+ template"
+    description: "Basic windows template long description with details"
+    openshift.io/long-description: >-
+      Long description of the template
+    openshift.io/provider-display-name: "Red Hat, Inc."
+    openshift.io/documentation-url: "https://cnv.io/..."
+    openshift.io/support-url: "https://access.redhat.com"
+    iconClass: icon-windows    
+
+    # The `defaults` set of annotations is meant as a hint only
+    # and is not going to be processed by the stock openshift templating
+    # mechanism. CNV UI can and should use it though.
+    # The information encoded there informs the tooling (UI)
+    # about names of devices that are to be used as templates
+    # for adding additional disk, networks, volumes and such
+
+    # The goal of default disk is to define what kind of disk
+    # is supported by the OS mainly in terms of bus (ide, scsi,
+    # sata, virtio, ...)
+    defaults.template.cnv.io/disk: default-disk
```

Comment:
1. Let's call it …/disk-bus
2. Let's use a sane example (i.e. virtio)

Reply:
Would this field always just contain one value (the bus)?

Reply:
No, it might contain anything that why it is a reference to a template disk.